lib/kube/proxy/transport.go

```diff
@@ -0,0 +1,419 @@
+// impersonation, a single transport is used per request. Otherwise, a new
+// transport is used for all requests in order to improve performance.
+// TODO(tigrato): Remove this once all servers support impersonation.
+func (f *Forwarder) transportForRequest(sess *clusterSession) (http.RoundTripper, error) {
+	// If all servers support impersonation, use a new transport for each
+	// request. This will ensure that the client certificate is valid for the
+	// server that the request is being sent to.
+	if !sess.teleportCluster.isRemote && allServersSupportImpersonation(sess.kubeServers) {
+		return f.transportForRequestWithImpersonation(sess)
+	}
+	// Otherwise, use a single transport per request.
+	return f.transportForRequestWithoutImpersonation(sess)
+}
+
+// DialContextFunc is a context network dialer function that returns a network connection
+type DialContextFunc func(context.Context, string, string) (net.Conn, error)
+
+// transportForRequestWithoutImpersonation returns a transport that does not
+// support impersonation. This is used when the cluster is remote or if the
+// servers don't support impersonation in order to ensure that the client request
+// can be routed to the correctly.
+//
+// DELETE in 14.0.0
+// TODO(tigrato): Remove this once all servers support impersonation.
+func (f *Forwarder) transportForRequestWithoutImpersonation(sess *clusterSession) (http.RoundTripper, error) {
+	if sess.creds != nil {
+		return sess.creds.getTransport(sess.upgradeToHTTP2), nil
+	}
+	transport := newTransport(sess.DialWithContext, sess.tlsConfig)
+	if !sess.upgradeToHTTP2 {
+		return tracehttp.NewTransport(transport), nil
+	}
+	if err := http2.ConfigureTransport(transport); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return tracehttp.NewTransport(transport), nil
+}
+
+// transportForRequestWithImpersonation returns a transport that supports
+// impersonation. This is used when the cluster is local and all servers support
+// impersonation. This allows the client to reuse the same transport for all
+// requests to the cluster in order to improve performance.
+// The transport is cached in the forwarder so that it can be reused for future
+// requests. If the transport is not cached, a new one is created and cached.
+// The transport is cached using the cluster name as the key.
+func (f *Forwarder) transportForRequestWithImpersonation(sess *clusterSession) (http.RoundTripper, error) {
+	// If the cluster is remote, the key is the teleport cluster name.
+	// If the cluster is local, the key is the teleport cluster name and the kubernetes
+	// cluster name: <teleport-cluster-name>/<kubernetes-cluster-name>.
+	key := transportCacheKey(sess)
+
+	// Check if the transport is cached.
+	cachedI, ok := f.cachedTransport.Get(key)
+	if ok {
+		if cached, ok := cachedI.(*httpTransport); ok {
+			if sess.upgradeToHTTP2 {
+				return cached.h2Transport, nil
+			}
+			return cached.h1Transport, nil
```

Comment:
`sess.creds`, now renamed to `sess.kubeAPICreds`, are the credentials to authenticate into the real Kubernetes API and are nil for all service that does not serve the Kubernetes cluster.

This means that `kube_proxy` and legacy `kube_proxy` can have the value nil because they won't serve the Kubernetes cluster.

I plan to refactor this code for Teleport 14 when we no longer need the old dialing method and I can safely remove a lot of legacy code that introduces a lot of complexity.